File: src/macos/AzPin/Services/ARMService.swift
import Foundation

protocol ARMServiceProtocol: Sendable {
    func fetchResourceGroups(subscriptionId: String, tenantId: String) async throws -> [AzureResourceGroup]
    func fetchResources(subscriptionId: String, resourceGroup: String, tenantId: String) async throws -> [AzureResource]
    func fetchAppState(resource: PinnedResource) async throws -> AppRunningState
    func startApp(resource: PinnedResource) async throws
    func stopApp(resource: PinnedResource) async throws
    func restartApp(resource: PinnedResource) async throws
}

extension ARMServiceProtocol {
    func fetchResourceGroups(subscriptionId: String) async throws -> [AzureResourceGroup] {
        try await fetchResourceGroups(subscriptionId: subscriptionId, tenantId: "")
    }

    func fetchResources(subscriptionId: String, resourceGroup: String) async throws -> [AzureResource] {
        try await fetchResources(subscriptionId: subscriptionId, resourceGroup: resourceGroup, tenantId: "")
    }
}

enum AppRunningState: Sendable {
    case running, stopped, unknown, starting, stopping, restarting
}

@Observable
final class ARMService: ARMServiceProtocol {
    private let tokenCache: any TokenCacheProtocol
    private let session: URLSession

    init(tokenCache: any TokenCacheProtocol, session: URLSession = .shared) {
        self.tokenCache = tokenCache
        self.session = session
    }

    func fetchResourceGroups(subscriptionId: String, tenantId: String) async throws -> [AzureResourceGroup] {
        let token = try await tokenCache.token(for: subscriptionId, tenantId: tenantId)
        let url = URL(string: "https://management.azure.com/subscriptions/\(subscriptionId)/resourcegroups?api-version=2021-04-01")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(ResourceGroupListResponse.self, from: data).value
    }

    func fetchResources(subscriptionId: String, resourceGroup: String, tenantId: String) async throws -> [AzureResource] {
        let token = try await tokenCache.token(for: subscriptionId, tenantId: tenantId)
        let url = URL(string: "https://management.azure.com/subscriptions/\(subscriptionId)/resourceGroups/\(resourceGroup)/resources?api-version=2021-04-01")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, _) = try await session.data(for: request)
        return try JSONDecoder().decode(ResourceListResponse.self, from: data).value
    }

    func fetchAppState(resource: PinnedResource) async throws -> AppRunningState {
        let token = try await tokenCache.token(for: resource.subscriptionId)
        let apiVer = Self.apiVersion(for: resource.type)
        let url = URL(string: "https://management.azure.com\(resource.id)?api-version=\(apiVer)")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, _) = try await session.data(for: request)
        return try Self.parseAppState(from: data, resourceType: resource.type)
    }

    func startApp(resource: PinnedResource) async throws {
        try await performAction("start", on: resource)
    }

    func stopApp(resource: PinnedResource) async throws {
        try await performAction("stop", on: resource)
    }

    func restartApp(resource: PinnedResource) async throws {
        // Container Apps have no restart endpoint; simulate with stop → start.
        if resource.type.lowercased() == "microsoft.app/containerapps" {
            try await performAction("stop", on: resource)
            try await performAction("start", on: resource)
        } else {
            try await performAction("restart", on: resource)
        }
    }

    private func performAction(_ action: String, on resource: PinnedResource) async throws {
        let token = try await tokenCache.token(for: resource.subscriptionId)
        let apiVer = Self.apiVersion(for: resource.type)
        let resolvedAction = Self.mapAction(action, for: resource.type)
        let url = URL(string: "https://management.azure.com\(resource.id)/\(resolvedAction)?api-version=\(apiVer)")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        _ = try await session.data(for: request)
    }

    static func apiVersion(for resourceType: String) -> String {
        switch resourceType.lowercased() {
        case "microsoft.app/containerapps": return "2023-05-01"
        case "microsoft.logic/workflows":   return "2019-05-01"
        default:                            return "2023-01-01"
        }
    }

    static func mapAction(_ action: String, for resourceType: String) -> String {
        guard resourceType.lowercased() == "microsoft.logic/workflows" else { return action }
        switch action {
        case "start": return "enable"
        case "stop":  return "disable"
        default:      return action
        }
    }

    static func parseAppState(from data: Data, resourceType: String) throws -> AppRunningState {
        switch resourceType.lowercased() {
        case "microsoft.app/containerapps":
            let response = try JSONDecoder().decode(ContainerAppResponse.self, from: data)
            return response.properties.runningStatus.lowercased() == "running" ? .running : .stopped
        case "microsoft.logic/workflows":
            let response = try JSONDecoder().decode(AppServiceResponse.self, from: data)
            return response.properties.state.lowercased() == "enabled" ? .running : .stopped
        default:
            let response = try JSONDecoder().decode(AppServiceResponse.self, from: data)
            return response.properties.state.lowercased() == "running" ? .running : .stopped
        }
    }
}

File: src/windows/AzPin.Windows.Core/Services/ArmService.cs
using System.Net.Http.Headers;
using System.Text.Json;
using AzPin.Windows.Models;
using AzPin.Windows.Models.Arm;

namespace AzPin.Windows.Services;

public class ArmException(string message, int statusCode, string body) : Exception(message)
{
    public int StatusCode { get; } = statusCode;
    public string Body { get; } = body;
}

public class ArmService(ITokenCache tokenCache, IHttpClientFactory httpClientFactory) : IArmService
{
    private readonly ITokenCache _tokenCache = tokenCache;
    private readonly HttpClient _httpClient = httpClientFactory.CreateClient("arm");
    private static readonly JsonSerializerOptions JsonOptions = new() { PropertyNameCaseInsensitive = true };

    private static readonly HashSet<string> RunnableTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "microsoft.web/sites",
        "microsoft.web/sites/slots",
        "microsoft.app/containerapps",
        "microsoft.logic/workflows"
    };

    public async Task<IReadOnlyList<ArmSubscription>> FetchSubscriptionsAsync(CancellationToken ct = default)
    {
        var response = await GetAsync<ArmListResponse<ArmSubscription>>("/subscriptions?api-version=2022-12-01", string.Empty, string.Empty, ct);
        return response.Value;
    }

    public async Task<IReadOnlyList<ArmResourceGroup>> FetchResourceGroupsAsync(string subscriptionId, string tenantId, CancellationToken ct = default)
    {
        var path = $"/subscriptions/{Uri.EscapeDataString(subscriptionId)}/resourcegroups?api-version=2021-04-01";
        var response = await GetAsync<ArmListResponse<ArmResourceGroup>>(path, subscriptionId, tenantId, ct);
        return response.Value;
    }

    public async Task<IReadOnlyList<ArmResource>> FetchResourcesAsync(string subscriptionId, string tenantId, string resourceGroupName, CancellationToken ct = default)
    {
        var path = $"/subscriptions/{Uri.EscapeDataString(subscriptionId)}/resourceGroups/{Uri.EscapeDataString(resourceGroupName)}/resources?api-version=2021-04-01";
        var response = await GetAsync<ArmListResponse<ArmResource>>(path, subscriptionId, tenantId, ct);
        return response.Value;
    }

    public async Task<AppRunningState> FetchRunningStateAsync(string subscriptionId, string tenantId, ArmResource resource, CancellationToken ct = default)
    {
        if (!RunnableTypes.Contains(resource.Type))
            return AppRunningState.Unknown;

        try
        {
            var (apiVersion, useRunningStatus, runningValue, stoppedValue) = StateConfig(resource.Type);
            var path = $"{resource.Id}?api-version={apiVersion}";
            var response = await GetAsync<ArmResourceStateResponse>(path, subscriptionId, tenantId, ct);

            var rawState = useRunningStatus
                ? response.Properties?.RunningStatus
                : response.Properties?.State;

            return rawState?.ToLowerInvariant() == runningValue
                ? AppRunningState.Running
                : rawState?.ToLowerInvariant() == stoppedValue
                    ? AppRunningState.Stopped
                    : AppRunningState.Unknown;
        }
        catch (ArmException)
        {
            return AppRunningState.Unknown;
        }
    }

    public Task StartResourceAsync(string subscriptionId, string tenantId, ArmResource resource, CancellationToken ct = default)
        => PostActionAsync(subscriptionId, tenantId, resource, "start", ct);

    public Task StopResourceAsync(string subscriptionId, string tenantId, ArmResource resource, CancellationToken ct = default)
        => PostActionAsync(subscriptionId, tenantId, resource, "stop", ct);

    public async Task RestartResourceAsync(string subscriptionId, string tenantId, ArmResource resource, CancellationToken ct = default)
    {
        bool needsStopStart = resource.Type.ToLowerInvariant() is
            "microsoft.app/containerapps" or "microsoft.logic/workflows";
        if (needsStopStart)
        {
            await PostActionAsync(subscriptionId, tenantId, resource, "stop", ct);
            await PostActionAsync(subscriptionId, tenantId, resource, "start", ct);
        }
        else
        {
            await PostActionAsync(subscriptionId, tenantId, resource, "restart", ct);
        }
    }

    // Returns (apiVersion, useRunningStatusField, runningValue, stoppedValue) all lowercased for comparison
    internal static (string ApiVersion, bool UseRunningStatus, string RunningValue, string StoppedValue) StateConfig(string resourceType) =>
        resourceType.ToLowerInvariant() switch
        {
            "microsoft.app/containerapps" => ("2023-05-01", true,  "running", "stopped"),
            "microsoft.logic/workflows"   => ("2019-05-01", false, "enabled", "disabled"),
            _                             => ("2023-01-01", false, "running", "stopped")
        };

    internal static string ApiVersion(string resourceType) => resourceType.ToLowerInvariant() switch
    {
        "microsoft.app/containerapps" => "2023-05-01",
        "microsoft.logic/workflows"   => "2019-05-01",
        _                             => "2023-01-01"
    };

    internal static string MapAction(string action, string resourceType) =>
        resourceType.ToLowerInvariant() == "microsoft.logic/workflows"
            ? action switch { "start" => "enable", "stop" => "disable", _ => action }
            : action;

    private async Task PostActionAsync(string subscriptionId, string tenantId, ArmResource resource, string action, CancellationToken ct)
    {
        var mappedAction = MapAction(action, resource.Type);
        var apiVersion = ApiVersion(resource.Type);
        var path = $"{resource.Id}/{mappedAction}?api-version={apiVersion}";
        using var request = await CreateRequestAsync(HttpMethod.Post, path, subscriptionId, tenantId, ct);
        request.Content = new StringContent(string.Empty);

        using var response = await _httpClient.SendAsync(request, ct);
        var body = await response.Content.ReadAsStringAsync(ct);
        if (!response.IsSuccessStatusCode)
            throw new ArmException($"ARM {action} failed for {resource.Id}", (int)response.StatusCode, body);
    }

    private async Task<T> GetAsync<T>(string path, string subscriptionId, string tenantId, CancellationToken ct)
    {
        using var request = await CreateRequestAsync(HttpMethod.Get, path, subscriptionId, tenantId, ct);
        using var response = await _httpClient.SendAsync(request, ct);
        var body = await response.Content.ReadAsStringAsync(ct);

        if (!response.IsSuccessStatusCode)
            throw new ArmException($"ARM GET failed for {path}", (int)response.StatusCode, body);

        var value = JsonSerializer.Deserialize<T>(body, JsonOptions);
        return value ?? throw new ArmException($"Failed to deserialize ARM response for {path}", (int)response.StatusCode, body);
    }

    private async Task<HttpRequestMessage> CreateRequestAsync(HttpMethod method, string path, string subscriptionId, string tenantId, CancellationToken ct)
    {
        var token = string.IsNullOrWhiteSpace(subscriptionId)
            ? string.Empty
            : await _tokenCache.GetTokenAsync(subscriptionId, tenantId, ct);

        var request = new HttpRequestMessage(method, path);
        if (!string.IsNullOrEmpty(token))
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", token);

        return request;
    }
}
